### PR TITLE
Introduction to Cryptography

### PR DESCRIPTION
**Pull request by Neurrone**
Tentitive outline:

By midterm: 

* symmetric key encryption
	* Stream ciphers
	* Block ciphers
		* modes of operation: ECB, CBC, CTR, GCM and common attacks

After midterm:

* asymmetric key encryption
	* RSA, PKI, https
* integrity:
	* hash functions
	* signitures
* interesting applications
* other resources: security now podcast, understanding cryptography book, awesome cryptography page

### CODE REVIEW

**Comment by ablyx** [comment]
I think you can reorder your work. Consider introducing stream ciphers, how they work, then talk about one-time-pad which sounds like a more specific type of stream cipher

**Comment by Neurrone** [comment]
@ablyx chronologically, the one-time pad was invented first, and its impracticality inspired stream ciphers. Its also usually taught before stream ciphers. Thought one-time pad would also be easier to explain/understand first, since its simpler - the keystream is from a truly random source, each bit xored with bits of the plaintext. Do you find the reverse order easier?

**Comment by ablyx** [comment]
Oh, I learnt stream ciphers first. But I think no matter the order, it would be good to explain the idea of a cryptosystem first. (Plaintext, ciphertext, key, encryption, decryption) [you can just give the main idea here]. Then talk about how the key is generated and used to encrypt and decrypt in stream cipher/one time pad. And I think one of the bigger limitation of the stream cipher is the periodicity of the stream.

File: contents/security/cryptography.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Introduction to Cryptography
+
+Authors: Dickson Tan
+
+# Overview
+
+Cryptography is the discipline of developing methods for secure communication. 
+Understanding cryptography wwill enable you to correctly use cryptography libraries and combine cryptographic primitives to match the security needs of your application.
```
Comment (MightyCupcakes) [inline contents/security/cryptography.md]: `wwill` -> `will`

File: contents/security/cryptography.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+Modern cryptography involves developing methods to achieve the following:
```
Comment (MightyCupcakes) [inline contents/security/cryptography.md]: `Modern cryptography techniques involves developing methods to achieve the following objectives` ?

File: contents/security/cryptography.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+During decryption, each ciphertext bit is xored with the corresponding keystream bit to produce the plaintext.
+Notice that encryption and decryption are the same operation; this is possible since xoring a bit x with another bit y twice recovers the original bit; (x xor y) xor y = x.
+
+The simplest stream cipher is the [one-time pad](https://en.wikipedia.org/wiki/One-time_pad). In this cipher, the keystream used is bits from a truly random source, and is also the key.
+It is the only known cipher that cannot be cracked, even if the attacker has infinite computing power.
+This property is known as [perfect secrecy](https://crypto.stackexchange.com/questions/3896/simply-put-what-does-perfect-secrecy-mean); the ciphertext gives no additional information about the plaintext, so knowing the ciphertext does not provide any advantage to the attacker trying to recover the plaintext.
+Unfortunately, Shannon proved that any cipher that achieves perfect secrecy has the following limitations, making them [impractical](https://www.schneier.com/crypto-gram/archives/2002/1015.html#7).
```
Comment (MightyCupcakes) [inline contents/security/cryptography.md]: Shannon? Might want a link to his page

File: contents/security/cryptography.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+[RC4](https://en.wikipedia.org/wiki/RC4) is the most widely used stream cipher. Though its use is now discouraged. The [eSTREAM project](http://www.ecrypt.eu.org/stream/) is a research effort to develop state-of-the-art stream ciphers.
+
+### Block Ciphers
+
+Unlike stream ciphers, which operate on individual bits, block ciphers operate on an entire block of bits at a time. In practice, the size of each block is 64 or 128 bits.
+
+Shannon introduced 2 primitives, which modern block ciphers are built on.
+
+* Confusion: an operation wwhich obscures the relationship between key and ciphertext. This is usually done by substitution.
```
Comment (MightyCupcakes) [inline contents/security/cryptography.md]: `wwhich` -> `which`

**Review by MightyCupcakes — COMMENTED** [review]
Overall good job in explaining a highly technical topic by abstracting the details away.

File: contents/security/cryptography.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+It is the only known cipher that cannot be cracked, even if the attacker has infinite computing power.
+This property is known as [perfect secrecy](https://crypto.stackexchange.com/questions/3896/simply-put-what-does-perfect-secrecy-mean); the ciphertext gives no additional information about the plaintext, so knowing the ciphertext does not provide any advantage to the attacker trying to recover the plaintext.
+Unfortunately, Shannon proved that any cipher that achieves perfect secrecy has the following limitations, making them [impractical](https://www.schneier.com/crypto-gram/archives/2002/1015.html#7).
+
+* The key must be truely random, not pseudorandomly generated, and must never be reused.
+* The key must be securely distributed, and be at least as long as the message being generated. For example, to send a 10gb file to someone encrypted with the one-time pad requires sending 10gb of key material + 10gb of ciphertext = 20gb.
+
+In practice, we do not require perfect secrecy, since attackers have limited computational power. Hence, all other ciphers are only secure against such attackers.
```
Comment (ablyx) [inline contents/security/cryptography.md]: try rephrasing "Hence, all other ciphers are only secure against such attackers." it's a bit confusing

File: contents/security/cryptography.md (hunk 3 above)
Comment (ablyx) [reply contents/security/cryptography.md]: maybe you should introduce Shannon

File: contents/security/cryptography.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+Modern stream ciphers approximate the operation of the one-time pad. 
+A short key (say 256 bits) is used to seed a [cryptographically secure pseudorandom number generator](https://en.wikipedia.org/wiki/Cryptographically_secure_pseudorandom_number_generator), which is used to generate the keystream for both encryption and decryption.
+They are more practical, as the key the communicating parties need to share is much shorter.
+
+Stream ciphers are used for their efficiency, ease of implementation in hardware, and when the length of the plaintext is unpredictable.
+However, block ciphers are more widely used than stream ciphers. In some modes of operation, they can be used like stream ciphers, reducing the need for dedicated stream ciphers.
+
+[RC4](https://en.wikipedia.org/wiki/RC4) is the most widely used stream cipher. Though its use is now discouraged. The [eSTREAM project](http://www.ecrypt.eu.org/stream/) is a research effort to develop state-of-the-art stream ciphers.
```
Comment (ablyx) [inline contents/security/cryptography.md]: why is usage of RC4 discouraged?

File: contents/security/cryptography.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+* Diffusion: an operation which hides statistical properties in the plaintext by spreading the influence of a plaintext bit over many ciphertext bits. For example, the DES cipher achieves this by bit Permutations.
+
+Ciphers that use only one of these operations are insecure. For example, the insecure Caesar cipher only uses confusion. But strong ciphers can be built by using both confusion and diffusion - these are called product ciphers.
+
+* This [article](https://graquantum.com/blog/deciphering-encryption-des-block-cipher/) explains how the  DES cipher works, Feistel networks, s-boxes and p-boxes. Though DES is no longer secure, its design has inspired many ciphers. A still secure variant, triple DES, is popular in legacy applications.
+* The Advanced Encryption Standard (AES) is the most popular symmetric cipher today. It is used by the US government, many protocols such as TLS, WPA2-AES and SSH. This [article](https://graquantum.com/blog/deciphering-encryption-aes-block-cipher/) explains how AES works without going too much into the mathematical details.
```
Comment (ablyx) [inline contents/security/cryptography.md]: `It is used by the US government, many protocols such as TLS, WPA2-AES and SSH` I dont understand the comma

**Comment by Neurrone** [comment]
Thanks @MightyCupcakes for the suggestions and catching the typos :)

**Review by ablyx — CHANGES_REQUESTED** [review]
some minor changes to be made

**Comment by Neurrone** [comment]
@ablyx changes made.

**Review by ablyx — APPROVED** [review]

**Review by MightyCupcakes — APPROVED** [review]

File: contents/security/cryptography.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+This [article](http://www.crypto-it.net/eng/theory/modes-of-block-ciphers.html) provides a nice overview of common modes. 
+
+Most modes require a random value called an initialization vector (IV) so that encrypting the same message twice doesn't produce the same ciphertext, which leaks information. 
+It is critical that the IV be [random, used only once and unpredictable](https://defuse.ca/cbcmodeiv.htm). Not doing so has caused several vulnerabilities such as the [BEAST Attack on TLS](http://www.educatedguesswork.org/2011/09/security_impact_of_the_rizzodu.html) and [the recovery of WEP keys](https://en.wikipedia.org/wiki/Wired_Equivalent_Privacy).
+
+## Other Resources
+
+* [Understanding Cryptography: A Textbook for Students and Practitioners](https://www.amazon.com/Understanding-Cryptography-Textbook-Students-Practitioners/dp/3642041000) is an outstanding introductory text. Explanations are excellent, and no knowledge of number theory is assumed. An electronic copy can be freely downloaded from the NUS library. It was used as reference material for this document.
```
Comment (damithc) [inline contents/security/cryptography.md]: Perhaps not refer to NUS as this is supposed to be a resource beyond NUS?

**Review by damithc — COMMENTED** [review]
Good work @Neurrone 
Some comments added. Revise as you see fit.
Adding more content and/or getting more peer reviews is welcome but not required. 
I will be merging end of week 12

**Comment by Neurrone** [comment]
@boxin-yang I've just added a short section on asymmetric cryptography and RSA.

**Comment by boxin-yang** [comment]
@Neurrone  Thx for the update. I will cite your book chapter as well as some other online sources :)

**Comment by damithc** [comment]
Merged via 97d8ca89ce3aa18b1629b63fec2f349d9eea697a

**Pull request closed without merging**